sys/mips/gt64120.c

```diff
@@ -466,6 +466,10 @@ pci_bus_driver_t gt_pci_bus = {
     .size = sizeof(gt_pci_state_t),
     .attach = gt_pci_attach,
     .probe = gt_pci_probe,
+    .interfaces = {
+      INTERFACE(struct pci_bus_driver, DEV_INTERFACE_BUS, bus),
```

Comment:
INTERFACE(struct bus_driver, DEV_INTERFACE_BUS, bus),

Reply:
I take it you suggest to change `struct pci_bus_driver` to `struct bus_driver`, right?
Why? With this new interface it's purely incidental that the offset of the `bus` field of `bus_driver` matches with with the one in `pci_bus_driver_t`.

Reply:
In my opinion it much better to change INTERFACE - to make it simpler and easier to read.
- the first what has been confusing (for me) is the argument order (I think the field like pci_bus, bus should be at the beginning) because other arguments are almost meaningless for reader.

from my point of view it would be much better when the driver interfaces are moved outside and sub-structs are just registered in ```.interface```
I mean something like
```
 bus_interface... bus_if = {
    .intr_setup = gt_pci_intr_setup,
    .intr_teardown = gt_pci_intr_teardown,
    .alloc_resource = gt_pci_alloc_resource,
    .release_resource = gt_pci_release_resource,
    .activate_resource = gt_pci_activate_resource,
    .deactivate_resource = gt_pci_deactivate_resource,
}

pci_interface... pci_bus_if = {
    .read_config = gt_pci_read_config,
    .write_config = gt_pci_write_config,
    .enable_busmaster = gt_pci_enable_busmaster,
}

pci_bus_driver_t gt_pci_bus = {
  .driver = {
    .desc = "GT-64120 PCI bus driver",
    .size = sizeof(gt_pci_state_t),
    .attach = gt_pci_attach,
    .probe = gt_pci_probe,
    .interfaces = {
      INTERFACE(DEV_INTERFACE_BUS, &bus_if),
      INTERFACE(DEV_INTERFACE_PCI, &pci_bus_if)
    }
}
```

imo it is better - because in this approach you drop all hard-to-read logic like ```offsetof```
because it is only philosophicalI - I don't insist. If you want to proceed with this approach I will give you aproval.
could you please let me know when this cs is ready for review. (or to label this PR as WIP)

Reply:
Maybe I don't see something obvious why it cannot be implemented in this way - please correct me if I am wrong

Reply:
The reason why I did it they way I did was because it's a slightly more general solution that works also for dynamically-allocated structures. I also liked having those interfaces in the same block and because at the time it was compatible with the inheritance-based solution (technically it still is in terms of structure layout but I removed sub-type definitions).

It's also _possibly_ slightly faster than using references, as all the data will most likely be in one block, however I doubt the compiler would put these structs far apart of each other and even if it did it would probably bear no real significance.